lib/actions/environment-variables.ts

```diff
@@ -0,0 +1,299 @@
+/*
+Copyright 2016-2017 Balena
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { CommandDefinition } from 'capitano';
+
+import * as commandOptions from './command-options';
+import { normalizeUuidProp } from '../utils/normalization';
+import { DeviceVariable, ApplicationVariable } from 'balena-sdk';
+
+const getReservedPrefixes = async (): Promise<string[]> => {
+	const balena = (await import('balena-sdk')).fromSharedOptions();
+	const settings = await balena.settings.getAll();
+
+	const response = await balena.request.send({
+		baseUrl: settings.apiUrl,
+		url: '/config/vars',
+	});
+
+	return response.body.reservedNamespaces;
+};
+
+export const list: CommandDefinition<
+	{},
+	{
+		application?: string;
+		device?: string;
+		config: boolean;
+	}
+> = {
+	signature: 'envs',
+	description: 'list all environment variables',
+	help: `\
+Use this command to list all environment variables for
+a particular application or device.
+
+This command lists all application/device environment variables.
+
+If you want to see config variables, used to configure
+balena features, use the --config option.
+
+At the moment the CLI does not support per-service variables,
+so the following commands will only show service-wide
+environment variables.
+
+Example:
+
+	$ balena envs --application MyApp
+	$ balena envs --application MyApp --config
+	$ balena envs --device 7cf02a6\
+`,
+	options: [
+		commandOptions.optionalApplication,
+		commandOptions.optionalDevice,
+
+		{
+			signature: 'config',
+			description: 'show config variables',
+			boolean: true,
+			alias: ['v', 'verbose'],
+		},
+	],
+	permission: 'user',
+	async action(_params, options, done) {
+		normalizeUuidProp(options, 'device');
+		const Bluebird = await import('bluebird');
+		const _ = await import('lodash');
+		const balena = (await import('balena-sdk')).fromSharedOptions();
+		const visuals = await import('resin-cli-visuals');
+
+		const { exitWithExpectedError } = await import('../utils/patterns');
+
+		return Bluebird.try(function(): Promise<
+			DeviceVariable[] | ApplicationVariable[]
+		> {
+			if (options.application != null) {
+				return balena.models.application[
+					options.config ? 'configVar' : 'envVar'
+				].getAllByApplication(options.application);
+			} else if (options.device != null) {
+				return balena.models.device[
+					options.config ? 'configVar' : 'envVar'
```

Comment:
I'm impressed that the typings can figure this out!
I would expect that we would first had to assign the functions to a variable with a ternary before invoking it.

Reply:
Yeah, I didn't think this would work either, but it totally does. TS is cool :sunglasses: